### ISSUE TITLE
Read/write .mif images and .tck files from python

### ISSUE THREAD

**Issue by dchristiaens** [MRtrix3/mrtrix3#1721]
This issue came up once again at the Workshop: it would be great to release code to read and write .mif and .tck files in Python, akin to the Matlab functions `read_mrtrix` etc.

I have code for this, which @maxpietsch has also expanded on, and I'm happy to release it but I don't know where to put it... 

- Do we make it part of the main repo? If so, do I put it in a separate folder (again akin two the Matlab support) or try integrate it with the `mrtrix3` scripting library?

- OR Should I keep it in a separate repo, set up as a full python package?

The main difference, I think, is that adding it to the scripting library means that users need to add the scripting library to their `PYTHONPATH`, whereas a separate package gives them the flexibility to decide if they need both. For instance, I have personally never felt the need to use the scripting libraries outside the scope of an MRtrix module, whereas with my IO module I have the opposite: I constantly use it in prototyping but never use it within an MRtrix module. Hence, the use cases for both seem to be quite different and I am therefore not yet convinced of the value of integrating both.

Happy to hear your thoughts...

**Comment by maxpietsch** [comment]
Love it! I'd suggest setting up a pip-compatible git repo for MRtrix file-IO and add it as a submodule to the mrtrix3 repo, probably best in `lib/mrtrix3/io`. This allows a standalone installation, lightweight integration into external (ML) applications and versioned usage within MRtrix3 (while slightly discouraging its usage over our C++ API). 

The only downside I see is that the submodule needs to be initialised if it ends up being used in the main repo, adding a potential source of installation error.

**Comment by Lestropie** [comment]
If use of such functionalities separately to the repo modules is common, then having them installable via `pip` and then importing them as a submodule if any future *MRtrix3* script requires them might be a good solution. Currently there's no need for the latter, but it would probably be worth putting the effort into testing to make sure that it *could* work. As long as the `build` script were to check the submodule it should be robust enough I think.

What would the module be called in that case? Are we going to run into naming issues trying to have the same code utilisable in both scenarios?

**Comment by dchristiaens** [comment]
> Love it! I'd suggest setting up a pip-compatible git repo for MRtrix file-IO and add it as a submodule to the mrtrix3 repo, probably best in `lib/mrtrix3/io`. This allows a standalone installation, lightweight integration into external (ML) applications and versioned usage within MRtrix3 (while slightly discouraging its usage over our C++ API).

Great! That ticks all the boxes. Love that it can be `pip` installable.

> The only downside I see is that the submodule needs to be initialised if it ends up being used in the main repo, adding a potential source of installation error.

Not too worried about this, since we'd want to discourage its use in the main scripting library anyways. If push comes to shove, we can always trigger the initialisation somewhere like we do in the `./run_tests` script.

> What would the module be called in that case? Are we going to run into naming issues trying to have the same code utilisable in both scenarios?

It would be great if we can name the python package `mrtrix3` and set  up a sub-package `mrtrix3.io`. Then users can do
```
from mrtrix3.io import read_mrtrix
```
without a direct naming clash with the scripting library. I also worry that this could trigger conflicts if the scripting library in in the Python path, but if it is correctly loaded as a `git` submodule I _think_ it could work. To be tested...

**Comment by dchristiaens** [comment]
OK, I have set up a new repo for the python package [here](https://github.com/dchristiaens/mrtrix3-pyio). It's currently under my name since I can't create new repositories within the MRtrix3 organisation, but this can be transferred easily. The structure is deliberately kept simple for now; I have for instance not yet added the `.tck` reader/writer.

I suggest we first test interoperability with the scripting library and the sub-repo mechanism that @maxpietsch suggested. Then we can look into registering on PyPI later.

**Comment by jdtournier** [comment]
> It's currently under my name since I can't create new repositories within the MRtrix3 organisation, but this can be transferred easily.

Let me know when you want me to create one for you. Not a problem.

> I suggest we first test interoperability with the scripting library and the sub-repo mechanism that @maxpietsch suggested. 

Given the discussion you've had about this, I'm not sold on the idea of making it a submodule - at least not at this stage. If the typical use case is more or less independent of the scripts we currently have or expect in the near future, then there is little point in making it a submodule. I think the `pip install` route would in practice be way more widely used. 

I guess what you're checking here is whether the structure can used in future scripts that would be distributed as part of the core _MRtrix3_ repo, right? So the test here is purely to ensure that if & when that happens, there will be no clashes? 

One thing I reckon might be problematic is that given the structure of your PIP repo, and the likely mount point of your submodule, you might have trouble achieving what you're after. I can't see any mechanism from the docs that would allow you to add the subfolder `mrtrix3/io` of your `mrtrix3-pyio` repo to the subfolder `lib/mrtrix3/io` of the main repo. From what I can tell, you can only add the _root_ of the submodule repo to the selected (empty) subfolder of the enclosing repo... (?)

**Comment by Lestropie** [comment]
> If the typical use case is more or less independent of the scripts we currently have or expect in the near future, then there is little point in making it a submodule. I think the pip install route would in practice be way more widely used.

It only takes one script relying on this IO to be contributed in the future, and it becomes a dependency; whether an explicit install through `pip` for all users (kind of unintuitive given the rest of the Python API is built-in), or a submodule, or some other solution. Given the likelihood of that happening eventually, I'd prefer to spend at least some time thinking about it now.

**Comment by maxpietsch** [comment]
Agree, probably mostly used via pip but the advantage of adding the IO repo as a submodule to mrtrix3 is that it is self-contained and versioned.

If `mrtrix3-pyio` is added as a submodule in the folder `lib/mrtrix3/io/mrtrix3-pyio`, it should be possible to incorporate it with a `__init__.py` file in `lib/mrtrix3/io`, no?
```
#__init__.py
from mrtrix3-pyio.mrtrix3.io import *
```
I think the idea is that within the MRtrix3 scripting library, one would use `mrtrix3.io` to get the submoduled version, if used outside the scripting library the same import would refer to the pip version.

**Comment by jdtournier** [comment]
> Given the likelihood of that happening eventually, I'd prefer to spend at least some time thinking about it now.

Makes sense. 

> If mrtrix3-pyio is added as a submodule in the folder lib/mrtrix3/io/mrtrix3-pyio, it should be possible to incorporate it with a __init__.py file in lib/mrtrix3/io, no?

Not a bad solution! OK, only remaining concern is how to handle submodule initialisation, especially for users either not installing from source, or installing without `git` from the tarball. 

I _think_ I'm starting to argue for including the module within the main _MRtrix3_ repo wholesale... I'm guessing it would still be possible to `pip install` the IO module if it's within the main repo?

**Comment by dchristiaens** [comment]
> It only takes one script relying on this IO to be contributed in the future, and it becomes a dependency;

For this reason, I'd caution against using it in the core scripts as long as possible. Don't forget that the IO package comes with a dependency on `numpy`, and may also be hard to maintain across python/numpy versions. (It has been tested with both python2 and python3, but depends on a recent numpy version to handle image strides correctly).

> Given the likelihood of that happening eventually, I'd prefer to spend at least some time thinking about it now.

I agree it's good to think about it now. If we would start using it, it would have to be versioned and the sub-repo mechanism looks very appealing for this. However...

> If `mrtrix3-pyio` is added as a submodule in the folder `lib/mrtrix3/io/mrtrix3-pyio`, it should be possible to incorporate it with a `__init__.py` file in `lib/mrtrix3/io`, no?

No, I don't think so. For this to work, `mrtrix3-pyio` needs to be an importable module, i.e., it needs to contain an `__init__.py` file. Yet, having such file in the top directory is discouraged in the recommended PyPI structure (I use [this](https://python-packaging.readthedocs.io/en/latest/minimal.html) template).

So I still don't see a clear path to a solution. I'm also still worried that, given the naming `mrtrix3` for both this python package and the scripting library, we are not tying ourselves into knots for the future...
